handler/file/retrieve.go

```diff
@@ -114,13 +114,88 @@ func (r *filecoinReader) Read(p []byte) (int, error) {
 		readLen = remainingBytes
 	}
 
+	buf := bytes.NewBuffer(p)
+	buf.Reset()
+
+	n, err := r.writeToN(buf, readLen)
+	return int(n), err
+}
+
+// WriteTo is implemented in order to directly handle io.Copy operations
+// rather than allow small, separate Read operations.
+func (r *filecoinReader) WriteTo(w io.Writer) (int64, error) {
+	if r.offset >= r.size {
+		return 0, io.EOF
+	}
+	// Read all remaining bytes and write them to w.
+	return r.writeToN(w, r.size-r.offset)
+}
+
+func (r *filecoinReader) writeToN(w io.Writer, readLen int64) (int64, error) {
```

Comment:
This is called by filecoinReader.Read(p []byte), which means it is called for every 32k data (that's the default behavior of io.Copy).

One issue caused by this is, if `findFileRanges` and `findProviders` is called for each Read(), that will have an impact on the overall performance if we are executing multiple SQL queries with JOIN for each 32k read.

In fact, the default size of each FileRange is 1GiB so it would be more efficient to find all file ranges and providers outside of Read() call.

Reply:
> Where is this used? I don't find io.WriterTo being used as filecoinReader
This is called within `io.Copy` if available, so that copying through the default buffer can be avoided, [here](https://cs.opensource.google/go/go/+/refs/tags/go1.21.3:src/io/io.go;l=409-413).

Reply:
> This is called by filecoinReader.Read(p []byte), which means it is called for every 32k data (that's the default behavior of io.Copy).
One issue caused by this is, if findFileRanges and findProviders is called for each Read(), that will have an impact on the overall performance if we are executing multiple SQL queries with JOIN for each 32k read.

Yes, this is designed to be called by for every 32k read, and specifically avoids calling `findFileRanges` and `findProviders` for each read. It does this by continuing to read from the read pipe previously set up to read from the range when it was first found.

The unit tests specifically force this situation and demonstrate how it is handled as a continuous read without dublication any calls to findFileRanges and findProviders, for the entirety of the range represented in the database.  

> In fact, the default size of each FileRange is 1GiB so it would be more efficient to find all file ranges and providers outside of Read() call.

If using the ranges provided by the outside request, then it would be possible for findFileRanges to be called multiple times for the same retrieval if an outside range happens to straddle multipiple ranges within the database. The code in the PR prevents that so that a retrieval is never duplicated, regardless of whether the requested ranges are withing the ranges in the database, cross multiple ranges within the database, or both.
The code in this PR continues reading from an established read pipe. If the